gemd/entity/object/has_template_check_generator.py

```diff
@@ -0,0 +1,71 @@
+"""For entities that have specs."""
+from gemd.entity.template.base_template import BaseTemplate
+from gemd.entity.link_by_uid import LinkByUID
+from gemd.entity.bounds_validation import get_validation_level, WarningLevel
+from gemd.entity.dict_serializable import logger
+
+from abc import ABC, abstractmethod
+from inspect import getmodule, getmembers, isclass, signature
+from typing import Union, Callable, TypeVar
+
+T = TypeVar('T')
+
+
+class HasTemplateCheckGenerator(ABC):
+    """Mix-in trait for objects that can generate a Check Template closure."""
+
+    @property
+    @abstractmethod
+    def template(self) -> Union[BaseTemplate, LinkByUID]:
+        """Get the object template associated with this object."""
+
+    def _generate_template_check(self,
+                                 validate: Callable[["HasTemplateCheckGenerator", T], bool]
+                                 ) -> Callable[[T], None]:
+        """
+        Generate a closure for the object and the validation routine.
+
+        Parameters
+        ----------
+        validate: function(HasTemplateCheckGenerator, Attribute) -> bool
+            A method that checks if the object template is consistent with the attribute
+
+        Returns
+        -------
+        function(Attribute)
+            If the attribute is inconsistent, ignores, warns or dies based on get_validation_level
+
+        Raises
+        ------
+        ValueError
+            If `value` is not one of the allowed types or if mapping types fails.
+
+        """
+        # Determine which class `validate` is from, so we can type check the object template
```

Comment:
I think there is a comment missing at the top of this block here that explains the problem that this module inspection stuff is solving.

As far as I understand it, you are passing a `validate` function which is referencing a method to some class. You need to check which module you are in and find the class type that is passing the function and use that class type in the closure you are generating. Is that correct? If so, at the top of this blob I think you should write clearly the problem this is solving. 

The inline comments help, but clearly stating the problem that this is addressing will make it easier to read for the next maintainer.